pkg/trace/writer/trace.go

```diff
@@ -283,19 +285,46 @@ func (w *TraceWriter) serializer() {
 			}
 
 			w.stats.BytesUncompressed.Add(int64(len(b)))
-			p := newPayload(map[string]string{
-				"Content-Type":     "application/x-protobuf",
-				"Content-Encoding": "zstd",
-				headerLanguages:    strings.Join(info.Languages(), "|"),
-			})
-			p.body.Grow(len(b) / 2)
-			zstdw := zstd.NewWriterLevel(p.body, zstd.BestSpeed)
-			if _, err := zstdw.Write(b); err != nil {
-				log.Errorf("Error zstd compressing trace payload: %v", err)
-			}
-			if err := zstdw.Close(); err != nil {
-				log.Errorf("Error closing zstd stream when writing trace payload: %v", err)
+			var p *payload
+
+			if w.useZstd {
+				p = newPayload(map[string]string{
+					"Content-Type":     "application/x-protobuf",
+					"Content-Encoding": "zstd",
+					headerLanguages:    strings.Join(info.Languages(), "|"),
+				})
+				p.body.Grow(len(b) / 2)
+
+				zstdw := zstd.NewWriterLevel(p.body, zstd.BestSpeed)
+				if _, err := zstdw.Write(b); err != nil {
+					log.Errorf("Error zstd compressing trace payload: %v", err)
+				}
+				if err := zstdw.Close(); err != nil {
+					log.Errorf("Error closing zstd stream when writing trace payload: %v", err)
+				}
+			} else {
+				p = newPayload(map[string]string{
+					"Content-Type":     "application/x-protobuf",
+					"Content-Encoding": "gzip",
+					headerLanguages:    strings.Join(info.Languages(), "|"),
+				})
+				p.body.Grow(len(b) / 2)
+
+				gzipw, err := gzip.NewWriterLevel(p.body, gzip.BestSpeed)
+				if err != nil {
+					// it will never happen, unless an invalid compression is chosen;
+					// we know gzip.BestSpeed is valid.
+					log.Errorf("gzip.NewWriterLevel: %d", err)
+					return
+				}
+				if _, err := gzipw.Write(b); err != nil {
+					log.Errorf("Error gzipping trace payload: %v", err)
+				}
+				if err := gzipw.Close(); err != nil {
+					log.Errorf("Error closing gzip stream when writing trace payload: %v", err)
+				}
```

Comment:
These cases can probably be de-duplacated a bit, as both `gzip.NewWriterLevel` and `zstd.NewWriterLevel` both return `io.WriteCloser` instances.

So something like
```
var (
    contentEncoding string
    w io.WriteCloser
)
if w.useZstd {
    contentEncoding = "zstd"
    w, err = zstd.NewWriterLevel(p.body, zstd.BestSpeed)
} else {
    contentEncoding = "gzip"
    w, err = gzip.NewWriterLevel(p.body, gzip.BestSpeed)
}
...
```

Also, even though the error should never occur, the error message
`log.Errorf("gzip.NewWriterLevel: %d", err)` is not useful to a user. If we ever change the code such that this log could occur, we want it to say something useful.

Reply:
Fixed thanks a lot.